python/visualization.py

```diff
@@ -0,0 +1,1076 @@
+    return ax
+
+def plot_monitors(sim,ax,output_plane=None,labels=False,monitor_parameters=None):
+    from meep.simulation import Volume
+
+     # consolidate plotting parameters
+    monitor_parameters = default_monitor_parameters if monitor_parameters is None else dict(default_monitor_parameters, **monitor_parameters)
+
+    label = 'monitor' if labels else None
+
+    for mon in sim.dft_objects:
+        for reg in mon.regions:
+            vol = Volume(center=reg.center,size=reg.size)
+            ax = plot_volume(sim,ax,vol,output_plane,plotting_parameters=monitor_parameters,label=label)
+    return ax
+
+def plot_fields(sim,ax=None,fields=None,output_plane=None,field_parameters=None):
+
+    if fields is None:
+        return ax
+    
+    # user specifies a field component
+    if fields in [mp.Ex, mp.Ey, mp.Ez, mp.Hx, mp.Hy, mp.Hz]:
+        # Get domain measurements
+        if output_plane:
+            sim_center, sim_size = (output_plane.center, output_plane.size)
+        elif sim.output_volume:
+            sim_center, sim_size = mp.get_center_and_size(sim.output_volume)
+        else:
+            sim_center, sim_size = (sim.geometry_center, sim.cell_size)
+        
+        xmin = sim_center.x - sim_size.x/2
+        xmax = sim_center.x + sim_size.x/2
+        ymin = sim_center.y - sim_size.y/2
+        ymax = sim_center.y + sim_size.y/2
+        zmin = sim_center.z - sim_size.z/2
+        zmax = sim_center.z + sim_size.z/2
+
+        center = Vector3(sim_center.x,sim_center.y,sim_center.z)
+        cell_size = Vector3(sim_size.x,sim_size.y,sim_size.z)
+
+        if sim_size.x == 0:
+            # Plot y on x axis, z on y axis (YZ plane)
+            extent = [ymin,ymax,zmin,zmax]
+            xlabel = 'Y'
+            ylabel = 'Z'
+        elif sim_size.y == 0:
+            # Plot x on x axis, z on y axis (XZ plane)
+            extent = [xmin,xmax,zmin,zmax]
+            xlabel = 'X'
+            ylabel = 'Z'
+        elif sim_size.z == 0:
+            # Plot x on x axis, y on y axis (XY plane)
+            extent = [xmin,xmax,ymin,ymax]
+            xlabel = 'X'
+            ylabel = 'Y'
+        fields = sim.get_array(center=center, size=cell_size, component=fields)
+    else:
+        raise ValueError('Please specify a valid field component (mp.Ex, mp.Ey, ...')
+    
+    # Either plot the field, or return the array
+    if ax:
+        field_parameters = default_field_parameters if field_parameters is None else dict(default_field_parameters, **field_parameters)
+        ax.imshow(np.rot90(fields), extent=extent, **field_parameters)
+        return ax
+    else:
+        return np.rot90(fields)
+    return ax
+
+def plot2D(sim,ax=None, output_plane=None, fields=None, labels=False,
+            eps_parameters=None,boundary_parameters=None,
+            source_parameters=None,monitor_parameters=None,
+            field_parameters=None):
+    
+    if not mp.am_master():
+        return
+
```

Comment:
All processes must call `init_sim`. Moving this check below line 494 may fix the deadlock.

Reply:
I switched it, but still see deadlocks. I'm trying to use your `visualize_chunks()` function as a template now. For whatever reason, that function does not deadlock, even though it is pretty similar.

Reply:
Although after carefully debugging, it did get rid of _some_ deadlocks...